File: Programs/LC 165/LC_165_1.py
# Runtime: 32 ms
# Faster Than: 83.77%
# Memory Usage: 14 MB
# Less Than: 52.71%

import os
cwd = os.path.dirname(__file__)

def compareVersion(version1: str, version2: str) -> int:
    version1 = list(map(int, version1.strip().split('.')))
    version2 = list(map(int, version2.strip().split('.')))

    i, j, diff = 0, 0, 0
    lenV1, lenV2 = len(version1), len(version2)

    while (i < lenV1) and (j < lenV2): 
        if version1[i] > version2[j]: 
            diff += 1
        elif version1[i] < version2[j]:
            diff -= 1
        diff *= 10

        i += 1
        j += 1
    
    while i < lenV1: 
        if version1[i] > 0: 
            diff += 1
        diff *= 10
        i += 1
    
    while j < lenV2: 
        if version2[j] > 0: 
            diff -= 1
        diff *= 10
        j += 1
    
    if diff < 0: return -1
    elif diff > 0: return 1

    return 0


if __name__ == '__main__': 
    nInputs = 5

    for i in range(1, nInputs + 1): 
        fname = "input{number}.txt"
        fpath = os.path.join(cwd, fname.format(number = i))
        
        print("======== TestCase {number} ========".format(number = i))
        try: 
            with open(fpath, "r") as fileHandle: 
                #  ===== START =====
                
                version1 = fileHandle.readline().strip()
                version2 = fileHandle.readline().strip()

                print(compareVersion(version1, version2))

                #  ===== END =====
        except FileNotFoundError:
            print("Cannot Open the test case")
            break

File: Programs/LC 165/LC_165_1.cpp
// Runtime: 3ms
// Faster Than: 31.35%
// Memory Usage: 6.2 MB
// Less Than: 35.61%

#include<iostream>
#include<fstream>
#include<vector>

using namespace std;

int compareVersion(string version1, string version2){
    int i = 0; 
    int j = 0; 

    int lenV1 = version1.length(); 
    int lenV2 = version2.length(); 

    int release = 0; 
    vector<int> releases1, releases2; 

    while(i < lenV1){
        if(version1[i] == '.'){
            releases1.push_back(release);             
            release = 0; 
        }else{
            release = (release * 10) + (version1[i] - '0'); 
        }
        i++; 
    }
    releases1.push_back(release); 
    release = 0;


    while(j < lenV2){
        if(version2[j] == '.'){
            releases2.push_back(release);             
            release = 0; 
        }else{
            release = (release * 10) + (version2[j] - '0'); 
        }
        j++; 
    }
    releases2.push_back(release); 

    i = 0; 
    j = 0; 
    lenV1 = size(releases1); 
    lenV2 = size(releases2); 

    while((i < lenV1) && (j < lenV2)){
        if(releases1[i] > releases2[j])
            return 1; 
        else if(releases1[i] < releases2[j])
            return -1; 
        ++i; 
        ++j; 
    }

    while(i < lenV1){
        if(releases1[i++] > 0) return 1; 
    }

    while(j < lenV2){
        if(releases2[j++] > 0) return -1; 
    } 

    return 0; 
}

int main(){
    // change the number of test cases
    int nInputs = 6;

    // test case files should be input1.txt, input2.txt, ..., inputN.txt format
    for(int i = 1; i <= nInputs; i++){
        string filename = "input" + to_string(i) + ".txt"; 

        ifstream fin(filename); 
        if(!fin.is_open()){
            cout<<"Cannot Open the test case"<<endl;
            return 0; 
        }

        cout<<"======== TestCase "<<i<<" ========"<<endl;


        // Write input and output here
        // Instead of cin use fin

        // ===== START =====
        string version1; fin>>version1; 
        string version2; fin>>version2; 

        cout<<compareVersion(version1, version2)<<endl; 

        // ===== END =====
    }
    
    return 0; 
}